active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,713 @@
+
+main.bicepparam
+
+```bicep
+using 'main.bicep'
+
+compatibility stacks
+
+param boolParam = true
+// ...
+
+extensionConfigs {
+  k8s: {
+    ✅ kubeConfig: az.aksKubeConfig(...) // any data needed to get it. Overloads may depend on auth workflow.
+    ❌ kubeConfig: loadFileContent('...')
+    ❌ kubeConfig: 'inlinedSecret'
+    namespace: 'default'
+  }
+}
+```
+
+If extensions require a configuration, there must be a parameters file (Bicep or JSON) that provide the extension
+configuration. If it is not supplied, there should be appropriate error diagnostics. There will be no support for
+inlining extension configurations from CLI arguments.
+
+### The new design from the ARM template perspective
+
+The main Bicep deployment would compile to the ARM template following this paragraph. Only REP relevant data is shown.
+The "extensionConfigs" deployment property will be similar in style to how deployment parameters and type definitions
+are defined in deployment templates. Configuration object property values can be literal values such as strings,
+booleans, and objects. Values can also be directives, such as to fetch a value from an ARM resource API. A directive
+consists of one or more evaluation steps. The extensions property supplied to the deployment PUT will support language
+expressions that will be evaluated against the template defining the deployment. This allows for use of non-secure
+deployment parameters to fill in portions of the configuration (namespace, resource names). The Deployment engine will
+be responsible for validation of expressions to ensure secret values are not processed in them.
+
+main.json - The root deployment
+
+```json5
+{
+  // root deployment
+  "template": {
+    "resources": {
+      "extResources": {
+        // Nested deployment
+        "type": "Microsoft.Resources/deployments",
+        "apiVersion": "2022-09-01",
+        "name": "extResourcesDeploymentName",
+        "properties": {
+          // Language expressions can be used. Expressions are not allowed to reference secure parameters or provide data for config secrets in ways that put the secret security at risk.
+          "extensionConfigs": {
+            "k8s": {
+              "type": "object",
+              // allowed types: "object", "inherit" (more on this in a further section)
+              "value": {
+                "namespace": {
+                  "type": "string",
+                  "value": "[parameters('namespace')]"
+                },
+                "kubeConfig": {
+                  "type": "armApiCall",
+                  "apiReference": {
+                    "method": "GET",
+                    "resourceId": "[ /subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/... ]",
+                    "apiVersion": "2024-02-01",
+                    "path": "/listClusterAdminCredentials",
+                    "query": "[ ?a=1&b=2 ]",
+                    "valueJsonPath": "[ kubeconfig[0].value ]"
```

Comment:
This seems like an implementation detail to me that could be handled on the back end. Converting an expression to this object structure would require the same amount of work, regardless of whether it’s done in the Bicep compiler or the deployment engine. What’s the advantage of this approach compared to simply emitting a template expression as the value and processing the expressions within the deployment engine?

Reply:
I'm undecided on this. I think either option would work and it depends on the complexity of each implementation. I will be digging into implementation details soon in my branch.

Reply:
One reason for the API reference approach is it's similar to how key vault references are expressed so these API references could resolved in the same code path as key vault references are now. For the value language expression, I'm uncertain currently where this translation would take place as I need to spend time on that area of the code.

Reply:
We may be able to support both, but we'd need the API reference variant at the root level (in params file) to source the kubeconfig from an AKS cluster if we're not using Azure based auth.

Reply:
I spent more time tracing code and thinking about stacks vs non-stacks. We'll have to go with general language expressions.